File: libsignal-service/src/main/java/org/whispersystems/signalservice/internal/push/SubscriptionsConfiguration.java
package org.whispersystems.signalservice.internal.push;

import com.fasterxml.jackson.annotation.JsonProperty;

import org.whispersystems.signalservice.api.profiles.SignalServiceProfile;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Response JSON for a call to /v1/subscriptions/configuration
 */
public class SubscriptionsConfiguration {

  public static final String PAYPAL     = "PAYPAL";
  public static final String CARD       = "CARD";
  public static final String SEPA_DEBIT = "SEPA_DEBIT";
  public static final String IDEAL      = "IDEAL";

  public static final int               BOOST_LEVEL        = 1;
  public static final int               GIFT_LEVEL         = 100;
  public static final int               BACKUPS_LEVEL      = 201;
  public static final HashSet<Integer>  SUBSCRIPTION_LEVELS = new HashSet<>(Arrays.asList(500, 1000, 2000));

  @JsonProperty("currencies")
  private Map<String, CurrencyConfiguration> currencies;

  @JsonProperty("levels")
  private Map<Integer, LevelConfiguration> levels;

  @JsonProperty("sepaMaximumEuros")
  private BigDecimal sepaMaximumEuros;

  @JsonProperty("backup")
  private BackupConfiguration backupConfiguration;

  public static class CurrencyConfiguration {
    @JsonProperty("minimum")
    private BigDecimal minimum;

    @JsonProperty("oneTime")
    private Map<Integer, List<BigDecimal>> oneTime;

    @JsonProperty("subscription")
    private Map<Integer, BigDecimal> subscription;

    @JsonProperty("backupSubscription")
    private Map<Integer, BigDecimal> backupSubscription;

    @JsonProperty("supportedPaymentMethods")
    private Set<String> supportedPaymentMethods;

    public BigDecimal getMinimum() {
      return minimum;
    }

    public Map<Integer, List<BigDecimal>> getOneTime() {
      return oneTime;
    }

    public Map<Integer, BigDecimal> getSubscription() {
      return subscription;
    }

    public Map<Integer, BigDecimal> getBackupSubscription() {
      return backupSubscription;
    }

    public Set<String> getSupportedPaymentMethods() {
      return supportedPaymentMethods;
    }
  }

  public static class LevelConfiguration {

    @JsonProperty("badge")
    private SignalServiceProfile.Badge badge;

    public SignalServiceProfile.Badge getBadge() {
      return badge;
    }
  }

  public static class BackupConfiguration {
    @JsonProperty("levels")
    private Map<Integer, BackupLevelConfiguration> backupLevelConfigurationMap;

    @JsonProperty("backupFreeTierMediaDays")
    private int freeTierMediaDays;

    public Map<Integer, BackupLevelConfiguration> getBackupLevelConfigurationMap() {
      return backupLevelConfigurationMap;
    }

    public int getFreeTierMediaDays() {
      return freeTierMediaDays;
    }
  }

  public static class BackupLevelConfiguration {
    @JsonProperty("storageAllowanceBytes")
    private long storageAllowanceBytes;

    public long getStorageAllowanceBytes() {
      return storageAllowanceBytes;
    }
  }

  public Map<String, CurrencyConfiguration> getCurrencies() {
    return currencies;
  }

  public Map<Integer, LevelConfiguration> getLevels() {
    return levels;
  }

  public BigDecimal getSepaMaximumEuros() {
    return sepaMaximumEuros;
  }

  public BackupConfiguration getBackupConfiguration() {
    return backupConfiguration;
  }
}

File: lib/libsignal-service/src/main/java/org/whispersystems/signalservice/internal/push/SubscriptionsConfiguration.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.whispersystems.signalservice.internal.push

import com.fasterxml.jackson.annotation.JsonProperty
import org.whispersystems.signalservice.api.profiles.SignalServiceProfile
import java.math.BigDecimal

/**
 * Response JSON for a call to /v1/subscriptions/configuration
 */
data class SubscriptionsConfiguration(
  val currencies: Map<String, CurrencyConfiguration> = emptyMap(),
  val levels: Map<Int, LevelConfiguration> = emptyMap(),
  val sepaMaximumEuros: BigDecimal? = null,
  @JsonProperty("backup") val backupConfiguration: BackupConfiguration = BackupConfiguration()
) {

  data class CurrencyConfiguration(
    val minimum: BigDecimal = BigDecimal.ZERO,
    val oneTime: Map<Int, List<BigDecimal>> = emptyMap(),
    val subscription: Map<Int, BigDecimal> = emptyMap(),
    val backupSubscription: Map<Int, BigDecimal> = emptyMap(),
    val supportedPaymentMethods: Set<String> = emptySet()
  )

  data class LevelConfiguration(
    val badge: SignalServiceProfile.Badge? = null
  )

  data class BackupConfiguration(
    @JsonProperty("levels") val backupLevelConfigurationMap: Map<Int, BackupLevelConfiguration> = emptyMap(),
    val freeTierMediaDays: Int = 0
  )

  data class BackupLevelConfiguration(
    val storageAllowanceBytes: Long = 0,
    val playProductId: String = "",
    val mediaTtlDays: Long = 0
  )

  companion object {
    const val PAYPAL = "PAYPAL"
    const val CARD = "CARD"
    const val SEPA_DEBIT = "SEPA_DEBIT"
    const val IDEAL = "IDEAL"

    const val BOOST_LEVEL = 1
    const val GIFT_LEVEL = 100
    const val BACKUPS_LEVEL = 201

    @JvmField
    val SUBSCRIPTION_LEVELS = setOf(500, 1000, 2000)
  }
}
